components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/nats/StreamingListener.java

```diff
@@ -0,0 +1,163 @@
+        CountDownLatch latch = new CountDownLatch(1);
+
+        SubscriptionOptions.Builder subscriptionOptions = new SubscriptionOptions.Builder();
+
+        String durableName = natsProperties.getProperty(NatsConstants.NATS_STREAMING_DURABLE_NAME);
+        String queueGroup = natsProperties.getProperty(NatsConstants.NATS_STREAMING_QUEUE_GROUP);
+        boolean isManualAck = Boolean.parseBoolean(natsProperties.getProperty(NatsConstants.NATS_STREAMING_MANUAL_ACK));
+        String ackWait = natsProperties.getProperty(NatsConstants.NATS_STREAMING_ACK_WAIT);
+        String maxInFlight = natsProperties.getProperty(NatsConstants.NATS_STREAMING_MAX_IN_FLIGHT);
+        String subscriptionTimeout = natsProperties.getProperty(NatsConstants.NATS_STREAMING_SUBSCRIPTION_TIMEOUT);
+        String dispatcher = natsProperties.getProperty(NatsConstants.NATS_STREAMING_DISPATCHER);
+
+        if (StringUtils.isNotEmpty(durableName)) {
+            subscriptionOptions.durableName(durableName);
+        }
+
+        if (isManualAck) {
+            subscriptionOptions.manualAcks();
+        }
+
+        if (StringUtils.isNotEmpty(ackWait)) {
+            subscriptionOptions.ackWait(Duration.ofSeconds(Integer.parseInt(ackWait)));
+        }
+
+        if (StringUtils.isNotEmpty(maxInFlight)) {
+            subscriptionOptions.maxInFlight(Integer.parseInt(maxInFlight));
+        }
+
+        if (StringUtils.isNotEmpty(subscriptionTimeout)) {
+            subscriptionOptions.subscriptionTimeout(Duration.ofSeconds(Integer.parseInt(subscriptionTimeout)));
+        }
+
+        if (StringUtils.isNotEmpty(dispatcher)) {
+            subscriptionOptions.dispatcher(dispatcher);
+        }
+
+        if (StringUtils.isNotEmpty(queueGroup)) {
+            connection.subscribe(subject, queueGroup, natsMessage -> {
+                String message = new String(natsMessage.getData());
+                log.info("Message Received to NATS Inbound EP: " + message);
+                injectHandler.invoke(message.getBytes(), sequenceName, null, null);
+                latch.countDown();
+                manuallyAck(isManualAck, natsMessage);
+            }, subscriptionOptions.build());
+        } else {
+            connection.subscribe(subject, natsMessage -> {
+                String message = new String(natsMessage.getData());
+                log.info("Message Received to NATS Inbound EP: " + message);
+                injectHandler.invoke(message.getBytes(), sequenceName, null, null);
+                latch.countDown();
+                manuallyAck(isManualAck, natsMessage);
+            }, subscriptionOptions.build());
+        }
+        latch.await();
+    }
+
+    private void manuallyAck(boolean isManualAck, Message natsMessage) {
+        if (isManualAck) {
+            try {
+                natsMessage.ack();
+            } catch (IOException e) {
+                log.error("An error occurred while sending manual ack. ", e);
```

Comment:
We cannot do anything here. Is it possible to mention the message ID here? Let's also say message might get redelivered.